dandi/dandifs.py

```diff
@@ -0,0 +1,434 @@
+        fs.glob('**/anat/*.json')
+
+    """
+
+    def __init__(
+            self,
+            dandiset: Optional[Union[str, RemoteDandiset]] = None,
+            version: Optional[str] = None,
+            client: Optional[Union[str, DandiInstance, DandiAPIClient]] = None,
+            **http_kwargs):
+        """
+        Initialize a remote DANDI file system.
+
+        The file system can be initialized from:
+
+        - a `RemoteDandiset` instance; or
+        - the name of a dandiset [+ version]; and
+
+            - a DandiAPIClient instance; or
+            - a DandiInstance instance; or
+            - the name of a known DANDI instance; or
+            - the url of a DANDI server.
+
+        :param dandiset: A `RemoteDandiset`, or the identifier of a
+            dandiset (e.g., `'000026'`).
+        :param version: The version of the dandiset to query
+            (e.g., `'draft'`).
+        :param client: An instantiated `DandiInstance` (or its
+            identifier) or an instantiated `DandiAPIClient` (or its url).
+        :param http_kwargs: Any other parameters passed on to the
+            `fsspec.HTTPFileSystem` file system.
+
+        """
+        self._httpfs = HTTPFileSystem(**http_kwargs)
+        super().__init__()
+        if not isinstance(dandiset, RemoteDandiset):
+            if isinstance(client, str):
+                if not client.startswith('http'):
+                    client = get_instance(client)
+            if isinstance(client, DandiInstance):
+                client = DandiAPIClient.for_dandi_instance(client)
+            else:
+                client = DandiAPIClient(client)
+            if dandiset:
+                dandiset = client.get_dandiset(dandiset, version)
+        self._dandiset = dandiset
+        self._client = None if dandiset else client
+
+    # ------------------------------------------------------------------
+    #   DANDI-specific helpers
+    # ------------------------------------------------------------------
+
+    @property
+    def dandiset(self) -> Optional[RemoteDandiset]:
+        """Default dandiset, used to browse relative paths"""
+        return self._dandiset
+
+    @dandiset.setter
+    def dandiset(self, x: Optional[RemoteDandiset]):
+        if x:
+            self._client = None
+        elif self._dandiset:
+            self._client = self._dandiset.client
+        self._dandiset = x
+
+    @property
+    def client(self) -> DandiAPIClient:
+        """Current client"""
+        return self.dandiset.client if self.dandiset else self._client
+
+    @client.setter
+    def client(self, x: DandiAPIClient):
+        if self.dandiset:
+            raise ValueError(
+                'Cannot assign a DANDI client to a FileSystem that is '
+                'already linked to a dandiset. Unassign the dandiset '
+                'first by: fs.dandiset = None')
+        self._client = x
+
+    @classmethod
+    def for_url(cls, url: str) -> "RemoteDandiFileSystem":
```

Comment:
Adding `from __future__ import annotations` to the top of the file will let you do:

```suggestion
    def for_url(cls, url: str) -> RemoteDandiFileSystem:
```